### ISSUE TITLE
Changing project on device doesn't remove old modules

### ISSUE THREAD

**Issue by sammachin** [flowforge/flowforge-device-agent#27]
### Current Behavior

When I change the project that my device is running and deploy a new snapshot for that project Node-RED still picks up and tries to install modules from the previous project.
I think this is due to `/opt/flowforge-device/project/node_modules` not being wiped.

### Expected Behavior

Deploying a new snapshot should result in a clean install of only the modules in the snapshpot

### Steps To Reproduce

Build project with a custom node.
Deploy to device,

Build new project without custom node.
Deploy to device.

Original custom node is still started.

### Environment

- FlowForge version: 0.8.0
- Agent 0.2.0
- Node.js version: 16.16
- npm version:
- Platform/OS:
- Browser:

**Comment by knolleary** [comment]
Check what the package.json contains in the project snapshot on the device after changing project.

We rewrite that file whenever the snapshot changes and it should only include the modules that were installed in the project it is assigned to. When then run `npm install` which should take care of removing any modules that are no longer needed. Deleting the whole of `node_modules` will mean node-red has to be reinstalled every time as well - so the goal is to avoid that.

**Comment by hardillb** [comment]
Also what version of npm do you have installed?

**Comment by sammachin** [comment]
@ben NPM 8.11.0

The package.json only has the project nodes, the sense-hat node and node-red which is what I would expect for this project/snapshot. But as you can see when it starts up its saying missing node modules for morse and serial port which were in the previous project.
```
pi@sensehatpi:/opt/flowforge-device/project $ flowforge-device-agent 
[2022-08-12T14:00:20.433Z] INFO: FlowForge Device Agent
[2022-08-12T14:00:20.440Z] INFO: Version: 0.2.0
[2022-08-12T14:00:20.440Z] INFO: Device: nEjxwoxM91
[2022-08-12T14:00:20.441Z] INFO: ForgeURL: https://forge.flowforge.dev
[2022-08-12T14:00:20.460Z] INFO: Project: 10f621bc-88c1-4461-94f0-ca0ae81dff4e
[2022-08-12T14:00:20.461Z] INFO: Snapshot: awbNlYPn40
[2022-08-12T14:00:20.462Z] INFO: Settings: d5f76bbffc9252f34a7c24aa6f6d919d9de932981800f3eea51245dd8149dba6
[2022-08-12T14:00:20.700Z] INFO: Starting heartbeat thread. Interval: 30s
[2022-08-12T14:00:21.356Z] INFO: Starting project
[NR] 12 Aug 15:00:23 - [info] 
[NR] Welcome to Node-RED
[NR] ===================
[NR] 12 Aug 15:00:23 - [info] Node-RED version: v3.0.1
[NR] 12 Aug 15:00:23 - [info] Node.js  version: v16.16.0
[NR] 12 Aug 15:00:23 - [info] Linux 5.15.32-v7+ arm LE
[NR] 12 Aug 15:00:25 - [info] Loading palette nodes
[NR] 12 Aug 15:00:27 - [warn] Missing node modules:
[NR] 12 Aug 15:00:27 - [warn]  - node-red-node-serialport (1.0.1): serial-port, serial out, serial in, serial request
[NR] 12 Aug 15:00:27 - [warn]  - node-red-contrib-morse (0.0.9): morse-out
[NR] 12 Aug 15:00:27 - [info] Settings file  : /opt/flowforge-device/project/settings.js
[NR] 12 Aug 15:00:27 - [info] Context store  : 'default' [module=memory]
[NR] 12 Aug 15:00:27 - [info] Installing module: node-red-node-serialport, version: 1.0.1
[NR] 12 Aug 15:00:27 - [info] User directory : /opt/flowforge-device/project
[NR] 12 Aug 15:00:27 - [warn] Projects disabled : set editorTheme.projects.enabled=true to enable
[NR] 12 Aug 15:00:27 - [info] Flows file     : /opt/flowforge-device/project/flows.json
[NR] 12 Aug 15:00:27 - [info] Admin UI disabled
[NR] 12 Aug 15:00:27 - [info] Server now running at http://127.0.0.1:1880
[NR] 12 Aug 15:00:27 - [warn] Encrypted credentials not found
[NR] 12 Aug 15:00:27 - [info] Starting flows
[NR] 12 Aug 15:00:27 - [info] Started flows
[NR] 12 Aug 15:00:27 - [info] Project Link nodes connected
[NR] 12 Aug 15:00:33 - [warn] Installation of module node-red-node-serialport failed:
[NR] 12 Aug 15:00:33 - [warn] ------------------------------------------
[NR] 12 Aug 15:00:33 - [warn] npm WARN config production Use `--omit=dev` instead.
[NR] npm ERR! code 127
[NR] npm ERR! path /opt/flowforge-device/project/node_modules/@serialport/bindings-cpp
[NR] npm ERR! command failed
[NR] npm ERR! command sh -c node-gyp-build
[NR] npm ERR! sh: 1: node-gyp-build: not found
[NR] npm ERR! A complete log of this run can be found in:
[NR] npm ERR!     /home/pi/.npm/_logs/2022-08-12T14_00_29_081Z-debug-0.log
[NR] 12 Aug 15:00:33 - [warn] ------------------------------------------
```
Those modules are still present in node-modules which is where i assume Node-RED is finding them?

**Comment by hardillb** [comment]
OK, that's a different problem

That's Node-RED running the old flow not that the nodes are still installed.

**Comment by sammachin** [comment]
```
pi@sensehatpi:/opt/flowforge-device/project $ cat flows.json 
[{"id":"e0b2bb93cbbbdc6d","type":"tab","label":"Flow 1","disabled":false,"info":"","env":[]},{"id":"21df73199aa9763d","type":"project link in","z":"e0b2bb93cbbbdc6d","name":"project in 1","project":"all","broadcast":false,"topic":"device","x":160,"y":220,"wires":[["e42eab411f50f827"]]},{"id":"e42eab411f50f827","type":"debug","z":"e0b2bb93cbbbdc6d","name":"debug 1","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":400,"y":220,"wires":[]}]
```
The new flow only contains a link node and a debug node, thats all thats in this new project

**Comment by hardillb** [comment]
OK, sorry, wrong "missing nodes" warning

This means we need to delete the `.config.nodes.json` file between snapshots/projects

**Comment by knolleary** [comment]
I can see what is happning. This is not about running old flows.

**TL;DR**: we need to delete `.config.*` when changing project.

Node-RED keeps a record of what modules it has found in `.config.nodes.json`. When it starts up, it reloads that list and then checks to see if everything it saw last time is still there.

If anything is missing (ie, no longer in `node_modules`) it logs the `Missing node modules` message.

Then, if `autoInstall` is enabled, it tries to reinstall the missing modules. That is what you can see happening and why the modules 'reappear' in `node_modules`.

By deleting `.config.nodes.json` between snapshot refreshes, we remove Node-RED's memory of what was previously installed and will stop it trying to reinstall old thing.

**Comment by knolleary** [comment]
Darn - Ben beat me to it by seconds. If only I hadn't gone for the full explanation 😉

**Comment by hardillb** [comment]
Yes, it should do

**Comment by sammachin** [comment]
yes that did it, deleted both `.config.nodes.json` and `.config.nodes.json.backup`

New snapshot sent down and it started cleanly
